### PR TITLE
chore/test: fix and improve linting, type-checking and unit tests

### PR DESCRIPTION
**Pull request by gkalpak**
Fix linting and type-checking errors, fix broken unit tests and improve the associated configurations/setups.
Builds on top of @jzolnowski's f0d8064287b4fdad8a87d60ec4e6ab7e8d76b91f.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -50,16 +50,21 @@
     "eslint": "6.8.0",
     "jest": "29.4.2",
     "lint-staged": "13.1.2",
+    "memory-fs": "0.5.0",
     "prettier": "2.8.4",
     "typescript": "4.9.5"
   },
+  "optionalDependencies": {
+    "fast-crc32c": "1.0.7"
+  },
```
Comment (jzolnowski) [inline package.json:59]: Do we need this dependency? https://github.com/xlts-dev/verdaccio-google-cloud-storage/pull/17#discussion_r1128544600

File: src/storage.ts [hunk 2]
```diff
@@ -232,51 +223,57 @@ class GoogleCloudStorageHandler implements IPackageStorageManager {
...
-        } catch (err) {
-          this.logger.error(
-            { name: file.name, err: err.message },
-            'gcloud: check exist package @{name} has failed, cause: @{err}'
-          );
+  private async _fileExist(name: string, fileName: string): Promise<boolean> {
+    const file: File = this._buildFilePath(name, fileName);
+    try {
+      const data = await file.exists();
+      const exist = data[0];
```
Comment (jzolnowski) [inline src/storage.ts:264]: NIT
```suggestion
      const [exist] = data;
```

File: tests/datastore.spec.ts [hunk 3]
```diff
@@ -42,221 +45,180 @@ describe('Google Cloud Storage', () => {
...
-
-                    return Promise.resolve([[entity], {}]);
-                  },
-                };
-              }
-            },
-          };
+      test('should delete an entity', (done) => {
+        const cloudDatabase = getCloudDatabase(storageConfig);
+        // Mock the `Datastore#key()` method to be generate a fake key.
```
Comment (jzolnowski) [inline tests/datastore.spec.ts:126]:
```suggestion
        // Mock the `Datastore#key()` method to be generated a fake key.
```

**Review by jzolnowski — CHANGES_REQUESTED** [review]
Great work George! All the tests are passing 🎉 I wonder if we should define a `pre-commit` script where we would call the script `test`. This way we would be sure that with each commits the tests pass

File: package.json (hunk 1 above)
Comment (gkalpak) [reply package.json:59]: Nope :smiley:
I guess this is a git rebase issue. Removed.

File: src/storage.ts (hunk 2 above)
Comment (gkalpak) [reply src/storage.ts:264]: Not directly related to this PR, but changed to:

```diff
-      const data = await file.exists();
-      const exist = data[0];
+      const [exist] = await file.exists();
```

File: tests/datastore.spec.ts (hunk 3 above)
Comment (gkalpak) [reply tests/datastore.spec.ts]: Good catch :+1:
Changed to `to generate a fake key`.

**Comment by gkalpak** [comment]
Thx for the review, @jzolnowski :+1:
Ready for another round :smiley:

File: package.json [hunk 4]
```diff
@@ -50,16 +50,18 @@
     "eslint": "6.8.0",
     "jest": "29.4.2",
     "lint-staged": "13.1.2",
+    "memory-fs": "0.5.0",
     "prettier": "2.8.4",
     "typescript": "4.9.5"
   },
   "scripts": {
     "build": "npm run build:types && npm run build:js",
```
Comment (jzolnowski) [inline package.json:58]: "pre-commit": "npm run test",

**Comment by jzolnowski** [comment]
@gkalpak Thank you for making the changes quickly. A few more comments from my side. We're almost there 🎉

File: src/storage.ts [hunk 5]
```diff
@@ -179,35 +176,27 @@ class GoogleCloudStorageHandler implements IPackageStorageManager {
...
-            resumable: this.config?.bucketOptions?.resumable,
-          });
-          resolve(null);
-        } catch (err) {
-          reject(getInternalError(err.message));
-        }
-      }
-    );
+  private async _savePackage(name: string, metadata: Package): Promise<void> {
+    const file = this._buildFilePath(name, PACKAGE_JSON);
```
[the remark this answers is not in the document]
Comment (gkalpak) [reply src/storage.ts:180]: In this PR, I mainly aim to fix linting, type-checking and unit-testing.
I would like to refrain from other "general" refactorings/improvements (i.e. that are not specifically fixing a linting/type-checking/testing error).

Feel free to create a new PR with other refactorings/improvements if you feel inclined to :smiley:

File: src/storage.ts [hunk 6]
```diff
@@ -179,35 +176,27 @@ class GoogleCloudStorageHandler implements IPackageStorageManager {
...
+    try {
+      await file.save(this._convertToString(metadata), {
```
[the remark this answers is not in the document]
Comment (gkalpak) [reply src/storage.ts:182]: In this PR, I mainly aim to fix linting, type-checking and unit-testing.
I would like to refrain from other "general" refactorings/improvements (i.e. that are not specifically fixing a linting/type-checking/testing error).

Feel free to create a new PR with other refactorings/improvements if you feel inclined to :smiley:

File: src/storage.ts [hunk 7]
```diff
@@ -179,35 +176,27 @@ class GoogleCloudStorageHandler implements IPackageStorageManager {
...
-             */
-            resumable: this.config?.bucketOptions?.resumable,
-          });
-          resolve(null);
-        } catch (err) {
-          reject(getInternalError(err.message));
-        }
-      }
-    );
+  private async _savePackage(name: string, metadata: Package): Promise<void> {
```
[the remark this answers is not in the document]
Comment (gkalpak) [reply src/storage.ts:179]: I believe this (and other similarly named methods) are intentionally named likely this to convey their relation to their unprefixed counterparts. (I.e. the prefixed methods are called internally by the unprefixed methods.)

From a quick look, I don't see the purpose of having these prefixed methods at all, but I haven't fully tried to understand the architecture.

As mentioned in other comments, this is outside the scope of this PR, but feel free to create a new PR with refactorings/improvements.

File: src/storage.ts [hunk 8]
```diff
@@ -288,7 +283,7 @@ class GoogleCloudStorageHandler implements IPackageStorageManager {
             this.logger.debug({ url: this.name }, 'gcloud:  @{url} package already exists in the storage bucket');
             uploadStream.emit('error', packageAlreadyExist(name));
           } else {
-            const file =this._getBucket().file(`${this.name}/${name}`);
+            const file = this._getBucket().file(`${this.name}/${name}`);
```
[the remark this answers is not in the document]
Comment (gkalpak) [reply src/storage.ts:286]: In this PR, I mainly aim to fix linting, type-checking and unit-testing.
I would like to refrain from other "general" refactorings/improvements (i.e. that are not specifically fixing a linting/type-checking/testing error).

Feel free to create a new PR with other refactorings/improvements if you feel inclined to 😃

File: tests/storage.spec.ts [hunk 9]
```diff
@@ -2,12 +2,13 @@ import fs from 'fs';
 import path from 'path';
 import { Writable } from 'stream';
 
-import { DownloadResponse } from '@google-cloud/storage';
-import { IPackageStorageManager } from '@verdaccio/types';
-import { Logger, ILocalData, Callback, Package, IPackageStorage } from '@verdaccio/types';
+import { DownloadResponse, File } from '@google-cloud/storage';
 import { HTTP_STATUS, API_ERROR, VerdaccioError } from '@verdaccio/commons-api';
+import { Callback, ILocalData, IPackageStorage, IPackageStorageManager, Logger, Package } from '@verdaccio/types';
+import MemoryFileSystem from 'memory-fs';
```
[the remark this answers is not in the document]
Comment (gkalpak) [reply tests/storage.spec.ts:8]: Am i not? 😱

File: package.json (hunk 4 above)
Comment (gkalpak) [reply package.json:58]: In my experience, running all tests on each commit is bad DX.

I could entertain the idea of running them before pushing, but a much better option imo (more reliable, better DX) would be to set up CI on this repository. I have created #19 to track this.

Comment (jzolnowski) [reply package.json:58]: It's covered in separate [PR](https://github.com/xlts-dev/verdaccio-google-cloud-storage/pull/20)

File: src/storage.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (jzolnowski) [reply src/storage.ts:180]: Covered in a separate [PR](https://github.com/xlts-dev/verdaccio-google-cloud-storage/pull/21)

File: src/storage.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (jzolnowski) [reply src/storage.ts:182]: Covered in a separate [PR](https://github.com/xlts-dev/verdaccio-google-cloud-storage/pull/21)

File: src/storage.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (jzolnowski) [reply src/storage.ts:179]: There are a lot of methods with `_` as prefix. We would need to refactor most of the files to fix that. Let's leave this topic for now 😄

File: src/storage.ts [hunk 10]
```diff
@@ -179,35 +176,27 @@ class GoogleCloudStorageHandler implements IPackageStorageManager {
...
+      });
+    } catch (err) {
+      throw getInternalError(err.message);
+    }
   }
-  /* eslint-enable no-async-promise-executor */
 
   private _convertToString(value: Package): string {
     return JSON.stringify(value, null, '\t');
   }
```
[the remark this answers is not in the document]
Comment (jzolnowski) [reply src/storage.ts:198]: Covered in a separate [PR](https://github.com/xlts-dev/verdaccio-google-cloud-storage/pull/21)

File: src/storage.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (jzolnowski) [reply src/storage.ts:286]: Covered in a separate [PR](https://github.com/xlts-dev/verdaccio-google-cloud-storage/pull/21)

File: src/storage.ts [hunk 11]
```diff
@@ -288,7 +283,7 @@ class GoogleCloudStorageHandler implements IPackageStorageManager {
             this.logger.debug({ url: this.name }, 'gcloud:  @{url} package already exists in the storage bucket');
             uploadStream.emit('error', packageAlreadyExist(name));
           } else {
-            const file =this._getBucket().file(`${this.name}/${name}`);
+            const file = this._getBucket().file(`${this.name}/${name}`);
             this.logger.info({ url: file.name }, 'gcloud: the @{url} is being uploaded to the storage bucket');
             const fileStream = file.createWriteStream({
```
[the remark this answers is not in the document]
Comment (jzolnowski) [reply src/storage.ts:288]: Covered in a separate [PR](https://github.com/xlts-dev/verdaccio-google-cloud-storage/pull/21)

File: tests/storage.spec.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (jzolnowski) [reply tests/storage.spec.ts:8]: Sorry, my mistake. I was confused about the import `memory-fs` but we also sort by the type of import, and this one belongs to external resources.

**Review by jzolnowski — APPROVED** [review]
Most of the comments have been resolved in a separate [PR](https://github.com/xlts-dev/verdaccio-google-cloud-storage/pull/21).

The rest looks good. Well done George 🎉

File: tests/storage.spec.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (gkalpak) [reply tests/storage.spec.ts:8]: I want you to know that I take offence at the fact that you even thought I wasn't sorting imports alphabetically 😛

**Comment by gkalpak** [comment]
Awesome 🚀
Thx for the super-quick reviews 🐎

**Pull request merged**